Test launcher based on pytest (#366)
idefix-code/idefix: doc/source/testing/testLauncher.rst

==========================
Test launcher and reporter
==========================

Overview
--------

> The class :doc:`idfxTest <idfxTest>` provides the basement to implement an *Idefix* integration test for validation.

Reviewer: ```suggestion
The class :doc:`idfxTest <idfxTest>` provides the toolbox to implement an *Idefix* integration test for validation.
```

In order to ease launching all the tests, the user might prefer to use directly the ``./test.py`` command at the
root of the *Idefix* sources.

This script will run all the listed variants of *Idefix* and build a report in the terminal. At the
end of the run a standard ``junit.xml`` file is produced. This one can be translated into a browsable
HTML file.

Depencencies
------------

> Before using :doc:`idfxTest <idfxTest>` you need to install some Python Depencencies (possibly in a ``virtual env``):

Reviewer: ```suggestion
Before using :doc:`idfxTest <idfxTest>` you need to install some Python depencencies (possibly in a ``virtual env``):
```

    test.py::test_idefix_build_run_check[HD/sod-iso-idefix.ini-single-reconstruction-2] PASSED              [ 12%]
    test.py::test_idefix_build_run_check[HD/sod-iso-idefix-hll.ini-single-reconstruction-2] PASSED          [ 25%]
    test.py::test_idefix_build_run_check[HD/sod-iso-idefix-hllc.ini-single-reconstruction-2] PASSED         [ 37%]
    test.py::test_idefix_build_run_check[HD/sod-iso-idefix-tvdlf.ini-single-reconstruction-2] PASSED        [ 50%]
    test.py::test_idefix_build_run_check[HD/sod-idefix.ini-single-reconstruction-2] PASSED                  [ 62%]
    test.py::test_idefix_build_run_check[HD/sod-idefix-hll.ini-single-reconstruction-2] PASSED              [ 75%]
    test.py::test_idefix_build_run_check[HD/sod-idefix-hllc.ini-single-reconstruction-2] PASSED             [ 87%]
    test.py::test_idefix_build_run_check[HD/sod-idefix-tvdlf.ini-single-reconstruction-2] PASSED            [100%]

    ---------- generated xml file: idefix-tests.junit.xml ---------------------------------------------------------
    =============================== 8 passed, 44 deselected in 73.03s (0:01:13) ===================================

> When seeing an error, the output of the command will be printed at the end.

Reviewer: ```suggestion
When an error is detected, the output of the command will be printed at the end.
```

    -init                 Reinit reference files for non-regression tests (dangerous!)
    -Werror               Consider warnings as errors
    -ccache               Use ccache to reduce the build time over multiple run of the test suite.
    -restart              Enable creating a restart from a checkpoint.
    -v, --verbose         Enable verbose mode, by not capturing the output.
    --help-pytest         Display the options you can transmit directly to pytest in addition to the specific to idefix tests.
    -fake                 Make a fake run by just logging the actions to validate that we generate same command over refactoring.
    -subdir SUBDIR        Select the test in the given subdir not to run all.

The script is built on top of the ``pytest`` command so you automatically get access
to all the advanced option this command provide. Here a few examples :

.. code-block:: shell

    # get the pytest help
    ./test.py --help-pytest

    # let pytest filtering the tests
    ./test.py -k "single and mpi"

    # stop on first failure
    ./test.py -x

    # re-run only the last failed tests
    ./test.py --lf

Definition of the tests
-----------------------

> The script is simply searching all the files names ``testme.json`` into the ``test`` directory.

Reviewer: ```suggestion
The script looks for all the files named ``testme.json`` in the ``test`` directory and all its sub-directories.
```

> This file describe the combination of parameters to use to produce the list of *Idefix* run to perform.

Reviewer: ```suggestion
This file describes the combination of parameters used to produce the list of *Idefix* runs to check.
```

                "tolerance": 0
            },{
                "dumpname": "dump.0001.dmp",
                "ini": ["idefix-hll.ini"],
                "vectPot": true,
                "single": true,
                "reconstruction": 2,
                "mpi": [false, true],
                "standardTest": false,
                "tolerance": 0
            }
        ]
    }

* HD/sod-iso-idefix.ini
* HD/sod-iso-idefix.ini-mpi
* HD/sod-iso-idefix-hll.ini-single-vectPot
* HD/sod-iso-idefix-hll.ini-single-vectPot-mpi

Naming the test
---------------

If you prefer to see the options appearing in a specific order in the generated test name,
you can provide the key ``namings`` listing as a comma separated list the variables in
the order you want to see them composing the test name.

.. code-block:: json

    {
        "namings": "ini,single,mpi",
        "variants": {
            "dumpname": "dump.0001.dmp",
            "ini": ["idefix.ini","idefix-hll.ini"],
            "vectPot": false,
            "single": [false, true],
            "reconstruction": 2,
            "mpi": [false, true],
            "standardTest": false,
            "tolerance": 0
        }
    }

> By default, the alphabetical order will be used.

Reviewer: ```suggestion
By default, the alphabetical order is used.
```
